python/tvm/relay/ir_pass.py

```diff
@@ -1,16 +1,292 @@
+    This class can be extended by adding new members when more meta data is
+    needed.
+
+    Parameters
+    ----------
+    name : str
+        The pass name.
+
+    opt_level : int
+        The optimization level of this pass.
+
+    required : List[str]
+        The list of passes that are required by a certain pass.
+    """
+
+    def __init__(self, name, opt_level, required=None):
+        self.__init_handle_by_constructor__(_ir_pass.PassInfo, name, opt_level,
+                                            required)
+
+
+@register_relay_node
+class PassContext(RelayNode):
+    """The basis where a Relay optimization/analysis runs on.
+    Each pass context contains a number of auxiliary information that is used
+    to help an optimization pass. Such information includes the error reporter
+    to record the errors of during the optimization, etc.
+    """
+
+    def __init__(self):
+        self.__init_handle_by_constructor__(_ir_pass.PassContext)
+
+
+@register_relay_node
+class Pass(RelayNode):
+    """The base class of all passes. All methods here are just simple wrappers
+    that are implemented in the backend. They are defined for users to
+    conveniently interact with the base class.
+    """
+
+    def set_pass_context(self, pass_ctx):
+        """Setup the pass context for analysis and optimizations. This context
+        could be shared by different passes for sequential passes.
+
+        Parameters
+        ----------
+        pass_ctx : PassContext
+            The context that is used to help perform a certain pass or a series
+            of passes.
+        """
+        if not isinstance(pass_ctx, PassContext):
+            raise TypeError("pass_ctx is expected to be the PassContext type")
+        _ir_pass.SetContext(self, pass_ctx)
+
+    def get_pass_info(self):
+        """Get the pass meta."""
+        return _ir_pass.GetPassInfo(self)
+
+    def __call__(self, mod):
+        """Execute the pass. Note that for sequential pass, the dependency among
+        different passes will be resolved in the backend.
+
+        Parameters
+        ----------
+        mod : tvm.relay.Module
+            The module that a certain optimization is performed on.
+
+        Returns
+        -------
+        mod : tvm.relay.Module
+            The updated module after applying this pass.
+        """
+        return _ir_pass.RunPass(self, mod)
+
+
+@register_relay_node
+class ModulePass(Pass):
+    """A pass that works on tvm.relay.Module.
+
+    Parameters
+    ----------
+    name : str
+        The module pass name.
+
+    opt_level : int
+        The optimization level of this module pass.
+
+    required : List[str]
+        The list of passes that are required by a module pass.
+
+    pass_func : Callable[(tvm.relay.Module, PassContext) -> tvm.relay.Module]
+        The callback function that sketches a certain optimization.
+    """
+
+    def __init__(self, name, opt_level, pass_func, required=None):
```

Comment:
I am not sure if we really want to support constructor along with the construction function here. Perhaps we can just not define constructor and ask the user to use create the function

Reply:
I feel like it is probably necessary to leave it here so users have a flexibility to apply a pass_func directly without using decorators. I am okay with either way. I am remove it if you think it is unecessary